core/ipc/sink/off-heap/src/main/java/org/opennms/core/ipc/sink/offheap/QueueFileOffHeapDispatchQueueFactory.java

```diff
@@ -0,0 +1,117 @@
+ *******************************************************************************/
+
+package org.opennms.core.ipc.sink.offheap;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+import org.opennms.core.ipc.sink.api.AsyncPolicy;
+import org.opennms.core.ipc.sink.api.DispatchQueue;
+import org.opennms.core.ipc.sink.api.DispatchQueueFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class QueueFileOffHeapDispatchQueueFactory implements DispatchQueueFactory {
+
+    private static final Logger LOG = LoggerFactory.getLogger(QueueFileOffHeapDispatchQueueFactory.class);
+
+    private final int inMemoryEntrySize;
+    private final long offHeapSize;
+    private final int batchSize;
+    private final Path baseFilePath;
+
+    private final Map<String, DispatchQueue<?>> queues = new ConcurrentHashMap<>();
+
+    public QueueFileOffHeapDispatchQueueFactory(int inMemoryEntrySize, int batchSize, String offHeapSize,
+                                                String baseFilePath) {
+        this.inMemoryEntrySize = inMemoryEntrySize;
+        this.batchSize = batchSize;
+        this.offHeapSize = convertToBytes(offHeapSize);
+
+        if (baseFilePath == null || baseFilePath.length() == 0) {
+            this.baseFilePath = Paths.get(System.getProperty("karaf.data"));
+        } else {
+            this.baseFilePath = Paths.get(baseFilePath);
+        }
+
+        LOG.info("DispatchQueue factory initialized with on-heap size: {}, batch size: {}, off-heap size: {}, " +
+                        "and file path: {}", this.inMemoryEntrySize, this.batchSize, this.offHeapSize,
+                this.baseFilePath);
+    }
+
+    @Override
+    public <T> DispatchQueue<T> getQueue(AsyncPolicy asyncPolicy, String moduleName, Function<T, byte[]> serializer,
+                                         Function<byte[], T> deserializer) {
+        if (asyncPolicy.getNumThreads() > inMemoryEntrySize) {
+            throw new IllegalArgumentException("The in memory queue size must be greater than or equal to the number" +
+                    " of consuming threads");
+        }
+
+        return (DispatchQueue<T>) queues.computeIfAbsent(moduleName, (k) -> {
+            try {
+                // TODO: Configurable
+                return new QueueFileOffHeapDispatchQueue<>(serializer, deserializer, k, baseFilePath,
+                        inMemoryEntrySize, batchSize,
+                        offHeapSize);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
+            }
+        });
+    }
+
+    private static long convertToBytes(String sizeWithSuffix) {
+        if (sizeWithSuffix == null || sizeWithSuffix.length() == 0) {
+            return 0;
+        }
+
+        String suffix = sizeWithSuffix.substring(sizeWithSuffix.length() - 2).toLowerCase();
```

Comment:
Will cause error if length is 1

Reply:
Might be a good idea to trim trailing white-space off of this too

Reply:
Catching that error and logging now.